app/src/main/java/com/oracle/javafx/scenebuilder/app/welcomedialog/WelcomeDialogWindowController.java

```diff
@@ -215,17 +225,105 @@ private void openDocument() {
+        }
+        question.setCancelButtonTitle(I18N.getString("alert.welcome.file.not.found.no"));
+        question.setMessage(shortMessage.toString());
+        question.setDetails(withPath);
+        
+        return question;
+    }
+    
+    boolean filePathExists(String filePath) {
+        return Files.exists(Path.of(filePath));
+    }
 
-    private void handleOpen(List<String> paths) {
+    /**
+     * Attempts to open files in filePaths. Scene Builder will only attempt to load
+     * files which exist. If a file does not exist, Scene Builder will ask the user
+     * to remove this file from recent files.
+     * 
+     * @param filePaths List of file paths to project files to be opened by Scene
+     *                  Builder.
+     */
+    private void handleOpen(List<String> filePaths) {
+        handleOpen(filePaths, 
+                   this::askUserToRemoveMissingRecentFiles,
+                   this::attemptOpenExistingFiles);
+    }
+
+    private void askUserToRemoveMissingRecentFiles(List<String> missingFiles) {
+        if (!missingFiles.isEmpty()) {
+            var questionDialog = questionMissingFilesCleanup(getStage(), missingFiles);
+            if (questionDialog.showAndWait() == AlertDialog.ButtonID.OK) {
+                removeMissingFilesFromPrefs(missingFiles);
+                loadAndPopulateRecentItemsInBackground();
+            }
+        }
+    }
+
+    private void attemptOpenExistingFiles(List<String> paths) {
         if (sceneBuilderApp.startupTasksFinishedBinding().get()) {
-            sceneBuilderApp.handleOpenFilesAction(paths);
-            getStage().hide();
+            openFilesAndHideStage(paths);
         } else {
-            showMasker(() -> {
-                sceneBuilderApp.handleOpenFilesAction(paths);
-                getStage().hide();
-            });
+            showMasker(() -> openFilesAndHideStage(paths));
+        }
+    }
+
+    private void openFilesAndHideStage(List<String> files) {
+        sceneBuilderApp.handleOpenFilesAction(files, () -> getStage().hide());
+    }
+
+    /**
+     * Attempts to open files in filePaths.
+     * In case of files are missing, a special procedure is applied to handle missing files.
+     *  
+     * @param filePaths List of file paths to project files to be opened by Scene Builder.
+     * @param missingFilesHandler Determines how missing files are handled.
+     * @param fileLoader Determines how files are loaded.
+     */
+    void handleOpen(List<String> filePaths, 
+                    Consumer<List<String>> missingFilesHandler,
+                    Consumer<List<String>> fileLoader) {
+
+        if (filePaths.isEmpty()) {
+            return;
         }
+
+        var candidates = filePaths.stream()
+                                  .collect(Collectors.groupingBy(this::filePathExists));
+
+        List<String> missingFiles = candidates.getOrDefault(Boolean.FALSE, new ArrayList<>());

```

Comment:
All of this is fine, but a bit hard to read the intent. The simplest form is effectively:

```
files.forEach(file -> {
    if (file.exists()) {
        open(file);
    } else {
        showError(file);
    }
});
```

Is there a way we can simplify the code to this form?

Reply:
I've reworked the code accordingly.